### ISSUE TITLE
Rendering bug - Extra blank pages are inserted depending on the content

### ISSUE THREAD

**Issue by guicara** [spipu/html2pdf#247]
Hello,

First of all, thank you for this library and the time spend maintaining it. 

I don't know if it is an issue with html2pdf or TCPDF.

Depending on the content of my PDF file, extra blank pages are inserted before and after the content. The number of blank pages inserted could be different if I change the content.

I was able to replicate this bug on a very simple PHP file. That's way you can perform tests and maybe you will understand why this bug occurs. For example you can try to add or remove paragraphs.

I use the master branch of html2pdf.

---

For the following examples, I output the PDF with this simple PHP code:

```
<?php
/**
 * Html2Pdf Library - example
 *
 * HTML => PDF converter
 * distributed under the OSL-3.0 License
 *
 * @package   Html2pdf
 * @author    Laurent MINGUET <webmaster@html2pdf.fr>
 * @copyright 2017 Laurent MINGUET
 */
require_once dirname(__FILE__).'/../vendor/autoload.php';

use Spipu\Html2Pdf\Html2Pdf;
use Spipu\Html2Pdf\Exception\Html2PdfException;
use Spipu\Html2Pdf\Exception\ExceptionFormatter;

try {
    ob_start();
    include dirname(__FILE__).'/content.php';
    $content = ob_get_clean();

    $html2pdf = new Html2Pdf('P', 'A4', 'fr');
    $html2pdf->setTestTdInOnePage(false);
    $html2pdf->writeHTML($content);
    $html2pdf->output('foo.pdf');
} catch (Html2PdfException $e) {
    $formatter = new ExceptionFormatter($e);
    echo $formatter->getHtmlMessage();
}
```

---

**Example n°1:**

Problem: 2 extra pages before and after the content.

![blank-1](https://user-images.githubusercontent.com/2470367/32503496-9d1dcaea-c3dd-11e7-9284-71f5a9b90dde.jpg)

HTML _(content.php)_:

```
<style type="text/css">
    <!--
    .article .entry-header .entry-title {
        margin: 0;
        line-height: 50px;
        font-size: 42px;
    }

    .article .entry-content .entry-title h1 {
        margin: 0;
        font-size: 42px;
    }

    .article .entry-content .entry-image {
        height: 220px;
        width: 670px;
    }

    .article .entry-content p {
        margin: 0 0 15px;
        font-size: 15px;
        line-height: 1.4;
    }

    -->
</style>
<page pageset="new" backtop="25mm" backbottom="38mm" backleft="11mm" backright="11mm">
    <page_header>
        header
    </page_header>
    <page_footer>
        footer
    </page_footer>
    <bookmark title="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut" level="1"></bookmark>
    <table id="post-185" class="article" cellspacing="0">
        <tr>
            <td>
                <div class="entry-content">
                    <div class="entry-title">
                        <h1>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut</h1>
                    </div>
                    <br/>
                    <br/>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                </div>
            </td>
        </tr>
    </table>
</page>
```

You can download the PDF here: 
[bug-example1.pdf](https://github.com/spipu/html2pdf/files/1450701/bug-example1.pdf)

---

**Example n°2:**

I have just removed the last paragraph in the content.

Problem: none!

![blank-2](https://user-images.githubusercontent.com/2470367/32504003-05836a26-c3df-11e7-954d-13070cf453ed.jpg)

HTML _(content.php)_:

```
<style type="text/css">
    <!--
    .article .entry-header .entry-title {
        margin: 0;
        line-height: 50px;
        font-size: 42px;
    }

    .article .entry-content .entry-title h1 {
        margin: 0;
        font-size: 42px;
    }

    .article .entry-content .entry-image {
        height: 220px;
        width: 670px;
    }

    .article .entry-content p {
        margin: 0 0 15px;
        font-size: 15px;
        line-height: 1.4;
    }

    -->
</style>
<page pageset="new" backtop="25mm" backbottom="38mm" backleft="11mm" backright="11mm">
    <page_header>
        header
    </page_header>
    <page_footer>
        footer
    </page_footer>
    <bookmark title="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut" level="1"></bookmark>
    <table id="post-185" class="article" cellspacing="0">
        <tr>
            <td>
                <div class="entry-content">
                    <div class="entry-title">
                        <h1>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut</h1>
                    </div>
                    <br/>
                    <br/>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                    <p>
                        Elementum integer enim neque volutpat ac tincidunt vitae. Lacus suspendisse faucibus interdum posuere lorem ipsum dolor sit. Duis at consectetur lorem donec. Cras ornare arcu dui vivamus arcu felis bibendum.
                        Scelerisque fermentum dui faucibus in ornare quam viverra orci. Risus quis varius quam quisque id. Consequat interdum varius sit amet mattis vulputate. Habitasse platea dictumst quisque sagittis purus sit. Praesent
                        tristique magna sit amet purus gravida quis blandit. Nunc lobortis mattis aliquam faucibus purus in massa tempor.
                    </p>
                    <p>
                        Sed faucibus turpis in eu mi bibendum neque egestas. Eu non diam phasellus vestibulum lorem sed risus ultricies. Dignissim enim sit amet venenatis urna cursus. Sagittis id consectetur purus ut faucibus pulvinar
                        elementum. Tellus mauris a diam maecenas sed enim ut. Eget duis at tellus at urna condimentum mattis pellentesque id. Malesuada fames ac turpis egestas. Turpis egestas integer eget aliquet nibh praesent tristique
                        magna. Commodo odio aenean sed adipiscing diam. Morbi non arcu risus quis varius quam quisque id diam. Ligula ullamcorper malesuada proin libero nunc consequat interdum varius sit. Viverra accumsan in nisl nisi
                        scelerisque. Ipsum suspendisse ultrices gravida dictum fusce ut placerat orci nulla. Commodo sed egestas egestas fringilla. Et malesuada fames ac turpis. Aliquet nibh praesent tristique magna. Lacinia at quis risus
                        sed vulputate odio. Facilisis sed odio morbi quis commodo odio aenean sed adipiscing. Dignissim enim sit amet venenatis urna cursus eget nunc scelerisque.
                    </p>
                </div>
            </td>
        </tr>
    </table>
</page>
```

You can download the PDF here: 
[bug-example2.pdf](https://github.com/spipu/html2pdf/files/1450727/bug-example2.pdf)

---

This is really a serious issue for our users. We produce a PDF version of a web magazine with a lot of pages. Each article of the magazine is inserted between a `<page>` tag. With a average of 20 articles in a magazine with have about 10 extra blank / empty pages!

As for now we have to edit the generated PDF with Adobe Acrobate to remove the extra pages and changes the pages number in the pagination and the table of contents. It is definitely not user friendly!

I hope you can help me!

Thanks,

**Comment by rajnishcoder** [comment]
Got the same problem :(

**Comment by rajnishcoder** [comment]
**I figure it out.** 
Don't use the `div` tag inside `table` as a main container like you're using `<div class="entry-content">`
Use only `table` tags and `p` as much as possible like you're creating an Email template.

**Comment by spipu** [comment]
Hi,
when you use `$html2pdf->setTestTdInOnePage(false);` you can have lots of pbs, like it says in the doc : https://github.com/spipu/html2pdf/blob/master/doc/methods.md#can-not-split-td-content

> **WARNING**: If you disable this test, you may have big layouts problems on tables in multiple columns.
> Rather than disabling this test it is better to break the TD content in smaller ones.
> **We do not support the consequencies of disable this test.**

Having such a big table, with only one TD, and only one DIV is not a good solution...

You car direct put the content in the page, without any div, table, ...

**Comment by guicara** [comment]
@rajnishcoder Thank you. I have deleted all `div` tags. Few bugs are now resolved but not all.

@spipu Thank you for your answer. I was not aware that if I set the method `setTestTdInOnePage` to `true` you do not support any bugs. Unfortunately if I set this method to `false`, I am unable to generate the PDF file:

> Html2Pdf Error [4]
> The content of the TD tag does not fit on only one page

... and indeed in the above example the TD tag does not fit on only one page. Beacause this tag is the wrapper of the article content. 

Therefore I don't understand how can I use `html2pdf` to output a content like that. Doest it means we can't use `html2pdf` to output more than one page of content?

Thanks,
